File: sdk/python/dchat/blockchain/transaction.py
"""
Transaction types for blockchain operations
"""

from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Optional


class TransactionStatus(Enum):
    """Transaction status enum"""
    PENDING = "Pending"
    CONFIRMED = "Confirmed"
    FAILED = "Failed"
    TIMED_OUT = "TimedOut"


class ChannelVisibility(Enum):
    """Channel visibility enum"""
    PUBLIC = "Public"
    PRIVATE = "Private"
    TOKEN_GATED = "TokenGated"


@dataclass
class TransactionReceipt:
    """Transaction receipt with confirmation details"""
    tx_id: str
    tx_hash: str
    success: bool
    block_height: Optional[int] = None
    block_hash: Optional[str] = None
    timestamp: Optional[datetime] = None
    error: Optional[str] = None


@dataclass
class RegisterUserTx:
    """Register user transaction"""
    user_id: str
    username: str
    public_key: str
    timestamp: datetime
    initial_reputation: int = 100

    def to_dict(self) -> dict:
        return {
            "user_id": self.user_id,
            "username": self.username,
            "public_key": self.public_key,
            "timestamp": self.timestamp.isoformat(),
            "initial_reputation": self.initial_reputation,
        }


@dataclass
class SendDirectMessageTx:
    """Send direct message transaction"""
    message_id: str
    sender_id: str
    recipient_id: str
    content_hash: str
    payload_size: int
    timestamp: datetime
    relay_node_id: Optional[str] = None

    def to_dict(self) -> dict:
        data = {
            "message_id": self.message_id,
            "sender_id": self.sender_id,
            "recipient_id": self.recipient_id,
            "content_hash": self.content_hash,
            "payload_size": self.payload_size,
            "timestamp": self.timestamp.isoformat(),
        }
        if self.relay_node_id:
            data["relay_node_id"] = self.relay_node_id
        return data


@dataclass
class CreateChannelTx:
    """Create channel transaction"""
    channel_id: str
    name: str
    description: str
    creator_id: str
    visibility: ChannelVisibility
    timestamp: datetime
    token_requirement: Optional[str] = None

    def to_dict(self) -> dict:
        data = {
            "channel_id": self.channel_id,
            "name": self.name,
            "description": self.description,
            "creator_id": self.creator_id,
            "visibility": self.visibility.value,
            "timestamp": self.timestamp.isoformat(),
        }
        if self.token_requirement:
            data["token_requirement"] = self.token_requirement
        return data


@dataclass
class PostToChannelTx:
    """Post to channel transaction"""
    message_id: str
    channel_id: str
    sender_id: str
    content_hash: str
    payload_size: int
    timestamp: datetime

    def to_dict(self) -> dict:
        return {
            "message_id": self.message_id,
            "channel_id": self.channel_id,
            "sender_id": self.sender_id,
            "content_hash": self.content_hash,
            "payload_size": self.payload_size,
            "timestamp": self.timestamp.isoformat(),
        }

File: sdk/dart/lib/src/blockchain/transaction.dart
/// Transaction types and models for blockchain operations
library;



/// Transaction status enum
enum TransactionStatus {
  pending,
  confirmed,
  failed,
  timedOut,
}

/// Transaction receipt containing confirmation details
class TransactionReceipt {
  final String txId;
  final String txHash;
  final bool success;
  final int? blockHeight;
  final String? blockHash;
  final DateTime timestamp;
  final String? error;

  TransactionReceipt({
    required this.txId,
    required this.txHash,
    required this.success,
    this.blockHeight,
    this.blockHash,
    required this.timestamp,
    this.error,
  });

  factory TransactionReceipt.fromJson(Map<String, dynamic> json) {
    return TransactionReceipt(
      txId: json['tx_id'] as String,
      txHash: json['tx_hash'] as String,
      success: json['success'] as bool,
      blockHeight: json['block_height'] as int?,
      blockHash: json['block_hash'] as String?,
      timestamp: DateTime.parse(json['timestamp'] as String),
      error: json['error'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'tx_id': txId,
      'tx_hash': txHash,
      'success': success,
      if (blockHeight != null) 'block_height': blockHeight,
      if (blockHash != null) 'block_hash': blockHash,
      'timestamp': timestamp.toIso8601String(),
      if (error != null) 'error': error,
    };
  }
}

/// Register user transaction
class RegisterUserTx {
  final String userId;
  final String username;
  final String publicKey;
  final DateTime timestamp;
  final int initialReputation;

  RegisterUserTx({
    required this.userId,
    required this.username,
    required this.publicKey,
    required this.timestamp,
    this.initialReputation = 100,
  });

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'username': username,
      'public_key': publicKey,
      'timestamp': timestamp.toIso8601String(),
      'initial_reputation': initialReputation,
    };
  }
}

/// Send direct message transaction
class SendDirectMessageTx {
  final String messageId;
  final String senderId;
  final String recipientId;
  final String contentHash;
  final int payloadSize;
  final DateTime timestamp;
  final String? relayNodeId;

  SendDirectMessageTx({
    required this.messageId,
    required this.senderId,
    required this.recipientId,
    required this.contentHash,
    required this.payloadSize,
    required this.timestamp,
    this.relayNodeId,
  });

  Map<String, dynamic> toJson() {
    return {
      'message_id': messageId,
      'sender_id': senderId,
      'recipient_id': recipientId,
      'content_hash': contentHash,
      'payload_size': payloadSize,
      'timestamp': timestamp.toIso8601String(),
      if (relayNodeId != null) 'relay_node_id': relayNodeId,
    };
  }
}

/// Channel visibility enum
enum ChannelVisibility {
  public,
  private,
  tokenGated,
}

/// Create channel transaction
class CreateChannelTx {
  final String channelId;
  final String name;
  final String description;
  final String creatorId;
  final ChannelVisibility visibility;
  final DateTime timestamp;
  final String? tokenRequirement;

  CreateChannelTx({
    required this.channelId,
    required this.name,
    required this.description,
    required this.creatorId,
    this.visibility = ChannelVisibility.public,
    required this.timestamp,
    this.tokenRequirement,
  });

  Map<String, dynamic> toJson() {
    return {
      'channel_id': channelId,
      'name': name,
      'description': description,
      'creator_id': creatorId,
      'visibility': visibility.name,
      'timestamp': timestamp.toIso8601String(),
      if (tokenRequirement != null) 'token_requirement': tokenRequirement,
    };
  }
}

/// Post to channel transaction
class PostToChannelTx {
  final String messageId;
  final String channelId;
  final String senderId;
  final String contentHash;
  final int payloadSize;
  final DateTime timestamp;

  PostToChannelTx({
    required this.messageId,
    required this.channelId,
    required this.senderId,
    required this.contentHash,
    required this.payloadSize,
    required this.timestamp,
  });

  Map<String, dynamic> toJson() {
    return {
      'message_id': messageId,
      'channel_id': channelId,
      'sender_id': senderId,
      'content_hash': contentHash,
      'payload_size': payloadSize,
      'timestamp': timestamp.toIso8601String(),
    };
  }
}
